File: content/courses/parallel-computing-introduction/codes/twod.cxx
#include <iostream>

using namespace std;

int main (int argc, char *argv[]) {
  int i, j;

  int nrl=4;
  int ncl=4;

  int nrows=nrl+2;
  int ncols=ncl+2;
  double **u=new double*[nrows];  
  double *buf=new double[nrows];

  double *uptr=new double[(nrows)*(ncols)];

  for (i=0;i<nrows;++i,uptr+=ncols)
     u[i] = uptr;

  for (i=0;i<nrows;++i) {
     for (j=0;j<ncols;++j) {
        u[i][j]=i+j*2;
     }
  }

  for (i=0;i<nrows;++i) {
     buf[i]=u[i%ncols][0];
  }

  //Make sure it's right
  for (i=0;i<nrows;++i) {
     for (j=0;j<ncols;++j) {
        cout<<u[i][j]<<" "; 
     }
     cout<<"\n";
  }

  cout<<"Buffer\n";
  for (i=0;i<nrows;++i) {
     buf[i]=u[0][i*ncols];
     cout<<buf[i]<<" ";
  }
  cout<<"\n";

}

File: content/courses/parallel-computing-introduction/codes/twod.f90
program twod
  integer :: i, j;
  integer :: nrl, ncl, nrows, ncols
  double precision, allocatable, dimension(:,:) :: u
  double precision, allocatable, dimension(:)   :: buf

  nrl=4;
  ncl=4;

  nrows=nrl+2;
  ncols=ncl+2;
  allocate(u(0:nrows-1,0:ncols-1))
  allocate(buf(0:ncols-1))

  do i=0,nrows-1
     do j=0,ncols-1
        u(i,j)=i+j*2
     enddo
  enddo

  do i=0,nrows-1
     buf(i)=u(0,mod(i,nrows))
  enddo

  !Make sure it's right
  do i=0,nrows-1
     do j=0,ncols-1
        write(*,'(f0.1,a)',advance='no') u(i,j)," "
     enddo
     write(*,*)
  enddo

  write(*,*) "Buffer"
  do i=0,nrows-1
     write(*,'(f0.1,a)',advance='no') buf(i)," "
  enddo
  write(*,*)

end program


